(guide): Using Push Notifications with Firebase in an Ionic Angular 4 App

bryplano opened the pull request:
## What's Changed
* Adds a walkthrough for setting up a basic integration with Firebase & the Capacitor Push Notification API for Android and iOS
* Updated sidebar with link to guide

## To Do
- [x] Android push test screenshot needs to be redone
- [x] Address any feedback

dotNetkow commented on site/src/components/site-menu/site-menu.tsx:
```diff
@@ -87,6 +87,10 @@ export class SiteMenu {
           url: '/docs/guides/ionic-framework-app'
+        },
+        {
+          title: 'Firebase Push Notifications',
```
thinking ahead to the future, maybe change the title to "Push Notifications - Firebase" or "Push Notifications with Firebase"?

dotNetkow commented on site/docs-md/guides/firebase-push-notifications.md:
```diff
@@ -0,0 +1,368 @@
+# Using Push Notifications with Firebase in an Ionic Angular 4 App
+
```
Suggestion about evergreen content: It's been challenging for me (and likely others) to keep version numbers etc up to date over time. Even though it's a tiny update, the mental aspect and actual effort takes more time than you'd think.  so.... :)

Consider mentioning "Ionic 4" specifically a bit less. here could be:
Web Framework: Ionic 4 + Angular

as i've got in my guide. then "Ionic App" over "Ionic 4" because future versions (5 and beyond) aren't going to change much.

dotNetkow commented on site/docs-md/guides/firebase-push-notifications.md:
```diff
@@ -0,0 +1,368 @@
+
+## Add Capacitor
+
+Once the application has been created successfully, switch to the newly created project directory. For example, on a Mac:
```
"on a Mac" -> strike this.  Changing directory (cd) is the same on Windows.

dotNetkow commented on site/docs-md/guides/firebase-push-notifications.md:
````diff
@@ -0,0 +1,368 @@
+const { PushNotifications } = Plugins;
+```
+
+Then, update the `ngOnInit()` method to register for push and `console.log()` a few of the events to listen for related events. My setup uses the `registration` / `registrationError` and `pushNotificationReceived` events:
````
No fancy Author component here unfortunately, so "my" might seem outta place.

dotNetkow commented on site/docs-md/guides/firebase-push-notifications.md:
````diff
@@ -0,0 +1,368 @@
+implementation 'com.google.firebase:firebase-messaging:17.4.0'
+```
+
+**Note**: Make sure that the file is added to source control. You may need to modify the `.gitignore` file under `android/` to add the file.
````
if this isn't fixed, can you let @jcesarmobile know?

bryplano replied:
Fixed: https://github.com/ionic-team/capacitor/pull/1495

dotNetkow commented on site/docs-md/guides/firebase-push-notifications.md:
```diff
@@ -0,0 +1,368 @@
+
+![Google Services JSON Location for Android](/assets/img/docs/guides/firebase-push-notifications/google-services-location-android.png)
+
+We don't need to *add* any permissions to our project because Capacitor projects automatically include a version of `firebase-messaging` in our `app/build.gradle` file:
```
great callout! definitely worth mentioning defaults, such as this case.

dotNetkow commented on site/docs-md/guides/firebase-push-notifications.md:
```diff
@@ -0,0 +1,368 @@
+
+### Prerequisites
+
+iOS push notifications are significantly more complicated to set up than Android. You must have a [paid Apple Developer account](https://developer.apple.com/) *and* take care of the following items prior to being able to test push notifications with your iOS application:
```
yes! must have an account is worth mentioning multiple times

dotNetkow commented on site/docs-md/guides/firebase-push-notifications.md:
```diff
@@ -0,0 +1,368 @@
+
+### Add the Firebase SDK via CocoaPods
+
+The Push Notification API on iOS makes use of CocoaPods - an iOS dependencies management system - and we need to tell CocoaPods to make use of Firebase.
```
nit: "dependencies" -> "dependency"

bryplano commented:
@dotNetkow addressed feedback. I do need to make one change to one of the screenshots tomorrow; otherwise, let me know if you have (or anyone else has) more feedback. :)

jcesarmobile commented on site/docs-md/guides/push-notifications-firebase.md:
```diff
@@ -0,0 +1,369 @@
+
+![Google Services JSON Location for Android](/assets/img/docs/guides/firebase-push-notifications/google-services-location-android.png)
+
+We don't need to *add* any permissions to our project because Capacitor projects automatically include a version of `firebase-messaging` in our `app/build.gradle` file:
```
push don't need permissions, also, you are talking about dependencies, not permissions.
Not sure if it's good to add this with versions as I'm working right now on updating to latest version (18.0.0), which will make this out of date shortly after being merged.

jcesarmobile commented on site/docs-md/guides/push-notifications-firebase.md:
```diff
@@ -0,0 +1,369 @@
+1. [Setup the proper Development or Production certificates & provisioning profiles](https://help.apple.com/xcode/mac/current/#/dev60b6fbbc7) for your iOS application in the Apple Developer Portal
+2. [Create an APNS certificate or key](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/establishing_a_certificate-based_connection_to_apns) for either Development or Production in the Apple Developer Portal
+3. [Ensure Push Notification capabilities have been enabled](https://help.apple.com/xcode/mac/current/#/dev88ff319e7) in your application in Xcode
+4. Have a physical iOS device running one of the two latest iOS versions (push notification functionality does not work in the iOS simulators)
```
Better say iOS 11 or newer, we expect to support iOS 11 when iOS 13 comes out at least for a while and that's 3 versions. Better update this to iOS 12 in the future if we drop iOS 11 support. Or link to the dependences page https://capacitor.ionicframework.com/docs/getting-started/dependencies (which I've noticed is wrong at the moment, will update it later)

jcesarmobile commented on site/docs-md/guides/push-notifications-firebase.md:
```diff
@@ -0,0 +1,369 @@
+
+![Google Service Info Plist Location for iOS](/assets/img/docs/guides/firebase-push-notifications/google-plist-location-ios.png)
+
+**Note**: It is recommended to use Xcode to move the `.plist` file into your project.
```
Not really sure what do you mean by "move the plist", move where?

jcesarmobile commented on site/docs-md/guides/push-notifications-firebase.md:
```diff
@@ -0,0 +1,369 @@
+
+The Push Notification API on iOS makes use of CocoaPods - an iOS dependency management system - and we need to tell CocoaPods to make use of Firebase.
+
+To do this, we need to modify the `Podfile`, which is located in your `ios/App` directory:
```
The Podfile is accessible from the Xcode project, probably better to open and edit it from there

jcesarmobile commented on site/docs-md/guides/push-notifications-firebase.md:
````diff
@@ -0,0 +1,369 @@
+We need to modify the App target pods to include Firebase. To do that, add `pod Firebase/Messaging` to your `target 'App'` section, like so:
+
+```ruby
+import Capacitor
````
this shouldn't be here, it's from the AppDelegate.swift, not from the Podfile

bryplano replied:
Oops, good find, this was meant to go in the `AppDelegate` code :)

jcesarmobile commented on site/docs-md/guides/push-notifications-firebase.md:
````diff
@@ -0,0 +1,369 @@
+```
+
+### Add Initialization Code
+
````
The `import Firebase` should be here in this section, not in the Podfile

jcesarmobile commented on site/docs-md/guides/push-notifications-firebase.md:
```diff
@@ -0,0 +1,369 @@
+
+### Update the Project
+
+Now we'll need to ensure that our iOS project is updated with the proper Firebase CocoaPod installed.
```
I think it's better to do this before doing the code changes as Firebase won't be available until this is run and will cause build failures on the Xcode project

jcesarmobile commented on site/docs-md/guides/push-notifications-firebase.md:
````diff
@@ -0,0 +1,369 @@
+npx cap open android
+```
+
+To open your iOS project in Xcode:
````
Also good to tell the users to open the project before doing the code changes you already told them to do.

jcesarmobile commented on site/docs-md/guides/push-notifications-firebase.md:
```diff
@@ -0,0 +1,369 @@
+
+Once the project is open, side-load the application on your device using the Run feature of either Android Studio or Xcode. The app should start up on the home page.
+
+**Note**: You may be asked to allow the application to receive notifications if this is the first time starting it up (this is the Push Notification API registering with Apple / Google). Make sure you choose to **Allow notifications**, or the next step won't work!
```
May no, you will for sure on iOS, and you won't on Android.

jcesarmobile commented on site/docs-md/guides/push-notifications-firebase.md:
```diff
@@ -0,0 +1,369 @@
+
+Now we'll need to ensure that our iOS project is updated with the proper Firebase CocoaPod installed.
+
+*Note*: This part can take awhile as CocoaPods needs to download all the appropriate files/dependencies.
```
a while instead of awhile

bryplano commented on site/docs-md/guides/push-notifications-firebase.md:
````diff
@@ -0,0 +1,369 @@
+Next, let's use the CLI to create a new Ionic 4 app based on the **blank** starter project and call it **capApp**:
+
+```bash
+ionic start capApp blank
````
Spoke to @dotNetkow on that, I have some ideas on how we can make these intro sections shorter, but we'll probably have to loop back to this. For now, planning to leave this as-is.
